File: app/src/main/java/de/lucaspape/monstercat/core/database/objects/Song.kt
package de.lucaspape.monstercat.core.database.objects

import android.content.Context
import android.net.Uri
import android.os.Build
import androidx.core.net.toUri
import com.google.android.exoplayer2.MediaItem
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory
import com.google.android.exoplayer2.upstream.DefaultHttpDataSource
import com.google.android.exoplayer2.util.Util
import de.lucaspape.monstercat.R
import de.lucaspape.monstercat.core.util.wifiConnected
import de.lucaspape.monstercat.core.util.Settings
import java.io.File

data class Song(
    val songId: String,
    private val title: String,
    private val version: String,
    val albumId: String,
    val mcAlbumId: String,
    val artist: String,
    val artistId: String,
    val isDownloadable: Boolean,
    private val isStreamable: Boolean,
    val inEarlyAccess: Boolean,
    val creatorFriendly: Boolean,
    val explicit: Boolean
) {

    companion object {
        @JvmStatic
        val TABLE_NAME = "song"

        @JvmStatic
        val COLUMN_SONG_ID = "songId"

        @JvmStatic
        val COLUMN_TITLE = "title"

        @JvmStatic
        val COLUMN_VERSION = "version"

        @JvmStatic
        val COLUMN_ALBUM_ID = "albumId"

        @JvmStatic
        val COLUMN_ALBUM_MC_ID = "mcALbumId"

        @JvmStatic
        val COLUMN_ARTIST = "artist"

        @JvmStatic
        val COLUMN_ARTIST_ID = "artistId"

        @JvmStatic
        val COLUMN_DOWNLOADABLE = "downloadable"

        @JvmStatic
        val COLUMN_STREAMABLE = "streamable"

        @JvmStatic
        val COLUMN_IN_EARLY_ACCESS = "inEarlyAccess"

        @JvmStatic
        val COLUMN_CREATOR_FRIENDLY = "creatorFriendly"

        @JvmStatic
        val COLUMN_EXPLICIT = "explicit"

        @JvmStatic
        val CREATE_TABLE =
            "CREATE TABLE " + TABLE_NAME + " (" +
                    COLUMN_SONG_ID + " TEXT PRIMARY KEY," +
                    COLUMN_TITLE + " TEXT," +
                    COLUMN_VERSION + " TEXT," +
                    COLUMN_ALBUM_ID + " TEXT," +
                    COLUMN_ALBUM_MC_ID + " TEXT," +
                    COLUMN_ARTIST + " TEXT," +
                    COLUMN_ARTIST_ID + " TEXT," +
                    COLUMN_DOWNLOADABLE + " TEXT," +
                    COLUMN_STREAMABLE + " TEXT," +
                    COLUMN_IN_EARLY_ACCESS + " TEXT," +
                    COLUMN_CREATOR_FRIENDLY + " TEXT," +
                    COLUMN_EXPLICIT + " TEXT" +
                    ")"
    }

    val shownTitle: String
        get() {
            return if (version.isNotEmpty()) {
                "$title ($version)"
            } else {
                title
            }
        }

    fun downloadLocation(context: Context): String {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            context.filesDir
                .toString() + "/" + artist + title + version + "." + Settings.getSettings(
                context
            ).getString("downloadType")
        } else {
            context.getExternalFilesDir(null)
                .toString() + "/" + artist + title + version + "." + Settings.getSettings(
                context
            ).getString("downloadType")
        }
    }

    fun downloadUrl(context: Context): String {
        return if (Settings.getSettings(context)
                .getBoolean(context.getString(R.string.useCustomApiForEverythingSetting)) == true && Settings.getSettings(
                context
            ).getBoolean(context.getString(R.string.customApiSupportsV1Setting)) == true
        ) {
            Settings.getSettings(context)
                .getString(context.getString(R.string.customDownloadUrlSetting)) + albumId + "/track-download/" + songId + "?format=" + Settings.getSettings(
                context
            ).getString("downloadType")
        } else {
            context.getString(R.string.trackContentUrl) + albumId + "/track-download/" + songId + "?format=" + Settings.getSettings(
                context
            ).getString("downloadType")
        }
    }

    fun deleteDownload(context: Context){
        File(downloadLocation(context)).delete()
    }

    private fun streamUrl(context: Context): String {
        return if (
            Settings.getSettings(context)
                .getBoolean(context.getString(R.string.useCustomApiForEverythingSetting)) == true && Settings.getSettings(
                context
            ).getBoolean(context.getString(R.string.customApiSupportsV1Setting)) == true
        ) {
            Settings.getSettings(context)
                .getString(context.getString(R.string.customStreamUrlSetting)) + albumId + "/track-stream/" + songId
        } else {
            context.getString(R.string.trackContentUrl) + albumId + "/track-stream/" + songId
        }
    }

    fun downloaded(context: Context): Boolean {
        return File(downloadLocation(context)).exists()
    }

    fun getMediaSource(
        context: Context,
        sid: String,
        cid: String,
        callback: (mediaSource: MediaSource?) -> Unit
    ) {
        if (downloaded(context)) {
            callback(fileToMediaSource(context, downloadLocation(context)))
        } else {
            if (isStreamable) {
                urlToMediaSource(context, streamUrl(context), sid, cid, callback)
            } else {
                callback(null)
            }
        }
    }

    fun playbackAllowed(context: Context): Boolean {
        return wifiConnected(context) || Settings.getSettings(
            context
        )
            .getBoolean(context.getString(R.string.streamOverMobileSetting)) == true || downloaded(
            context
        )
    }

    private fun fileToMediaSource(context: Context, fileLocation: String): ProgressiveMediaSource {
        return ProgressiveMediaSource.Factory(
            DefaultDataSourceFactory(
                context, Util.getUserAgent(
                    context, context.getString(R.string.app_name)
                )
            )
        ).createMediaSource(MediaItem.fromUri(Uri.parse("file://$fileLocation")))
    }

    private fun urlToMediaSource(
        context: Context,
        url: String,
        sid: String,
        cid: String,
        callback: (mediaSource: MediaSource) -> Unit
    ) {
        val httpSourceFactory =
            DefaultHttpDataSource.Factory()

        httpSourceFactory.setUserAgent(
            Util.getUserAgent(
                context,
                context.getString(R.string.app_name)
            )
        )

        var cookie = ""

        if (cid != "") {
            cookie += "cid=$cid"
        }

        if (sid != "") {
            cookie += ";connect.sid=$sid"
        }

        if (cookie.isNotEmpty()) {
            httpSourceFactory.setDefaultRequestProperties(mapOf("Cookie" to cookie))
        }

        callback(
            ProgressiveMediaSource.Factory(httpSourceFactory)
                .createMediaSource(MediaItem.fromUri(url.toUri()))
        )
    }
}

File: lib/objects/Song.dart
class Song {
  static const String TABLE_NAME = "song";
  static const String COLUMN_SONG_ID = "songId";
  static const String COLUMN_TITLE = "title";
  static const String COLUMN_VERSION = "version";
  static const String COLUMN_ALBUM_ID = "albumId";
  static const String COLUMN_ALBUM_MC_ID = "mcALbumId";
  static const String COLUMN_ARTIST = "artist";
  static const String COLUMN_ARTIST_ID = "artistId";
  static const String COLUMN_DOWNLOADABLE = "downloadable";
  static const String COLUMN_STREAMABLE = "streamable";
  static const String COLUMN_IN_EARLY_ACCESS = "inEarlyAccess";
  static const String COLUMN_CREATOR_FRIENDLY = "creatorFriendly";
  static const String CREATE_TABLE =
      "CREATE TABLE " + TABLE_NAME + " (" +
          COLUMN_SONG_ID + " TEXT PRIMARY KEY," +
          COLUMN_TITLE + " TEXT," +
          COLUMN_VERSION + " TEXT," +
          COLUMN_ALBUM_ID + " TEXT," +
          COLUMN_ALBUM_MC_ID + " TEXT," +
          COLUMN_ARTIST + " TEXT," +
          COLUMN_ARTIST_ID + " TEXT," +
          COLUMN_DOWNLOADABLE + " TEXT," +
          COLUMN_STREAMABLE + " TEXT," +
          COLUMN_IN_EARLY_ACCESS + " TEXT," +
          COLUMN_CREATOR_FRIENDLY + " TEXT" +
          ")";

  Map<String, String> toMap() {
    return <String, String>{
      COLUMN_SONG_ID: songId,
      COLUMN_TITLE: title,
      COLUMN_VERSION: version,
      COLUMN_ALBUM_ID: albumId,
      COLUMN_ALBUM_MC_ID: mcAlbumId,
      COLUMN_ARTIST: artist,
      COLUMN_ARTIST_ID: artistId,
      COLUMN_DOWNLOADABLE: isDownloadable.toString(),
      COLUMN_STREAMABLE: isStreamable.toString(),
      COLUMN_IN_EARLY_ACCESS: inEarlyAccess.toString(),
      COLUMN_CREATOR_FRIENDLY: isCreatorFriendly.toString()
    };
  }

  Song.fromMap(Map<String, String> map) {
    songId = map[COLUMN_SONG_ID];
    title = map[COLUMN_TITLE];
    version = map[COLUMN_VERSION];
    albumId = map[COLUMN_ALBUM_ID];
    mcAlbumId = map[COLUMN_ALBUM_MC_ID];
    artist = map[COLUMN_ARTIST];
    artistId = map[COLUMN_ARTIST_ID];
    isDownloadable = map[COLUMN_DOWNLOADABLE].toLowerCase() == 'true';
    isStreamable = map[COLUMN_STREAMABLE].toLowerCase() == 'true';
    inEarlyAccess = map[COLUMN_IN_EARLY_ACCESS].toLowerCase() == 'true';
    isCreatorFriendly = map[COLUMN_CREATOR_FRIENDLY].toLowerCase() == 'true';
  }

  String songId;
  String title;
  String version;
  String albumId;
  String mcAlbumId;
  String artist;
  String artistId;
  bool isDownloadable;
  bool isStreamable;
  bool inEarlyAccess;
  bool isCreatorFriendly;

  Song(
      {this.songId,
        this.title,
        this.version,
        this.albumId,
        this.mcAlbumId,
        this.artist,
        this.artistId,
        this.isDownloadable,
        this.isStreamable,
        this.inEarlyAccess,
        this.isCreatorFriendly});
}